### PR TITLE
Update to support new apis

### PR DESCRIPTION
**Pull request by shangardezi**
- WIP: upgrade library to support risk and filter apis
- Change track to filter
- Add missing arguments
- Tests
- Move context to first arg

### CODE REVIEW

File: castle.go [hunk 1]
```diff
@@ -9,29 +9,38 @@ import (
...
+	EventTypeProfileReset         EventType = "$profile_reset"
+	EventTypePasswordResetRequest EventType = "$password_reset_request"
+	EventTypeChallenge            EventType = "$challenge"
+)
+
+// EventStatus is an enum defining the statuses for a given event.
+type EventStatus string
 
-// See https://castle.io/docs/events
+// See https://docs.castle.io/docs/events
```
Comment (basia-horwath) [inline castle.go:39]: Not sure this doc helps with understanding the statuses

File: castle.go [hunk 2]
```diff
@@ -9,29 +9,38 @@ import (
...
-	EventPasswordResetRequestFailed    Event = "$password_reset_request.failed"
-	EventPasswordResetSucceeded        Event = "$password_reset.succeeded"
-	EventPasswordResetFailed           Event = "$password_reset.failed"
-	EventIncidentMitigated             Event = "$incident.mitigated"
-	EventReviewResolved                Event = "$review.resolved"
-	EventReviewEscalated               Event = "$review.escalated"
-	EventChallengeRequested            Event = "$challenge.requested"
-	EventChallengeSucceeded            Event = "$challenge.succeeded"
-	EventChallengeFailed               Event = "$challenge.failed"
+	EventStatusAttempted EventStatus = "$attempted"
```
Comment (basia-horwath) [inline castle.go:41]: Perhaps we should add `$requested` as well to enable people to use the endpoints with all the type?

File: castle.go [hunk 3]
```diff
@@ -199,30 +207,21 @@ func (c *Castle) SendTrackCall(e *castleAPIRequest) error {
 
 	resp := &castleAPIResponse{}
 
-	if resp.Error != "" {
+	if resp.Type != "" {
```
Comment (basia-horwath) [inline castle.go:210]: This still does nothing ;)

File: castle.go [hunk 4]
```diff
@@ -199,30 +207,21 @@ func (c *Castle) SendTrackCall(e *castleAPIRequest) error {
...
+		// we have an api error
+		return errors.New(resp.Type)
+	}
+
+	if resp.Message != "" {
```
Comment (basia-horwath) [inline castle.go:215]: And this neither

File: castle.go (hunk 1 above)
Comment (shangardezi) [reply castle.go:39]: In the events table it mentions the different statuses that are supported for events. 
e.g for $challenge:
```
A user is sent through an additional verification flow, such as Two-factor or Email-verification. This event supports the status $requested in addition to $failed and $succeeded.
```

Comment (basia-horwath) [reply castle.go:39]: Yeah, I know. But it does not list what statuses are available for what types.

File: castle.go [hunk 5]
```diff
@@ -137,47 +124,69 @@ func ContextFromRequest(r *http.Request) *Context {
...
 
-type castleAPIResponse struct {
-	Error       string `json:"error"`
-	Type        string `json:"type"`
-	Message     string `json:"message"`
-	Action      string `json:"action"`
-	UserID      string `json:"user_id"`
-	DeviceToken string `json:"device_token"`
+type User struct {
+	Id           string            `json:"id"`
```
Comment (basia-horwath) [inline castle.go:138]: Could this be `ID` instead of `Id`?

File: castle.go [hunk 6]
```diff
@@ -137,47 +124,69 @@ func ContextFromRequest(r *http.Request) *Context {
...
-func (c *Castle) Track(event Event, userID string, properties map[string]string, userTraits map[string]string, context *Context) error {
-	e := &castleAPIRequest{Event: event, UserID: userID, Context: context, Properties: properties, UserTraits: userTraits}
-	return c.SendTrackCall(e)
+type castleAPIResponse struct {
+	Type    string `json:"type"`
+	Message string `json:"message"`
+	Risk    string `json:"risk"`
+	Policy  struct {
+		Name       string `json:"name"`
+		Id         string `json:"id"`
```
Comment (basia-horwath) [inline castle.go:161]: Could this be ID instead of Id?

File: castle.go [hunk 7]
```diff
@@ -137,47 +124,69 @@ func ContextFromRequest(r *http.Request) *Context {
...
+		RevisionId string `json:"revision_id"`
```
Comment (basia-horwath) [inline castle.go:162]: Could this be `RevisionID` instead of `RevisionId`?

File: castle.go (hunk 3 above)
Comment (shangardezi) [reply castle.go:211]: good spot!

File: castle.go (hunk 1 above)
Comment (basia-horwath) [reply castle.go:39]: Maybe a link to the reference? (https://reference.castle.io/)

**Review by basia-horwath — APPROVED** [review]

**Pull request merged**